### ISSUE TITLE
complex_layout_scroll_perf__timeline_summary average_frame_rasterizer_time_millis regression

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#19434]
Starting with this engine roll: https://github.com/flutter/flutter/pull/19322

**Comment by cbracken** [comment]
From a quick look at the benchmarks, looks like about a ~7% regression on Android and a ~15% improvement on iOS for this benchmark. Equivalent benchmarks for flutter_gallery__transition_perf (iOS/Android) appear to be unaffected.

**Comment by cbracken** [comment]
@liyuqian this is one of the recent regressions that looks as though it may not be Dart 2 related. Taking a look at the engine commit list in #19277, most are related to Skia/compositing. Would you mind taking a look?

**Comment by cbracken** [comment]
See also #19433

**Comment by Hixie** [comment]
Update: We have determined that this is a Skia issue and are bisecting to find the exact commit. We will follow up with Skia once we know the cause.

**Comment by liyuqian** [comment]
Ignore the comment above as Skia is not causing the regression. I've been tricked by noise.

After more robust tests (running tests 10 times and takes the average instead of just looking at 1 test run), I believe that either the two Dart rolls, or the modification of flutter/flutter in https://github.com/flutter/flutter/pull/19322 are responsible for the regression according to the following experiments:

1. Checkout flutter/flutter to revision e22f99743b4f0bea4a9b8e7703ce5ffbfda37a04 right before https://github.com/flutter/flutter/pull/19322

2. Checkout flutter/engine to revision fed2ea458ed49088d33eddabc546ba56d600c717 which is the engine that https://github.com/flutter/flutter/pull/19322 rolled into.

3. Revert engine commits d9a831c5557ea06ff53260890d402b96d1af9470 and b157d4f5683023836cdf15d42bf56fd201a7084b, the two Dart rolls in the engine

4. Compile engine and run complex layout perf test against the local engine.

The perf does not regress in the experiments above, and the only difference between the regressed version and the above experiment is (1) two Dart rolls; (2) flutter/flutter changes in https://github.com/flutter/flutter/pull/19322

Hence, I'm changing the assignee to @aam 

**Note:** *please use Moto G4 for the test (I used Nexus 5X and the results are completely misleading), and please run the perf tests multiple times (at least 5, I used 10) and take the average before deciding whether there's a regression or not as the test is noisy and the regression is very small. It would have saved my 2 days of time if I followed these two strategies early on.*

**Comment by jason-simmons** [comment]
I can also reproduce this on a Moto G4.  Some notes:

* The benchmark's `99th_percentile_frame_rasterizer_time_millis` metric is a clearer indicator of whether this regression is occurring.
* It looks like this is associated with the version of the Dart SDK in the framework's `bin/cache/dart-sdk` directory.
    * I'm seeing this regression happen if I sync to the version of the framework prior to https://github.com/flutter/flutter/pull/19322 and replacing `bin/cache/dart-sdk` with the version included in the https://github.com/flutter/flutter/pull/19322 roll (2.0.0-edge.84ca27a09ebd6a65cd23ee52d835d89cbe06c574)
    * In the scenario above, note that the `libflutter.so` binary running on the device does not include the Dart roll.

**Comment by alexmarkov** [comment]
I was able to reproduce this regression as well. I used average of 10 runs, regression is about ~5-8%, with the noise up to 2-4%. Observations:

1) The regression does not happen when local engine is switched before and after the roll.

2) flutter drive does not use local engine to run scroll_perf_test.dart - it is run using downloaded dart-sdk even if --local-engine option is specified.

3) Regression appears if --preview-dart-2 is added to VM options which flutter drive uses to run scroll_perf_test.dart without any other changes (same dart-sdk before roll, same local engine). Code at https://github.com/flutter/flutter/blob/da0e898da9c329192cd1ef4f8fe6e7d03a952adc/packages/flutter_tools/lib/src/commands/drive.dart#L288-L295 has a bug: it passes '--preview-dart-2' after main Dart script, so it has no effect.

4) The roll which causes regression switches downloaded dart-sdk to dart 2 mode by default.

5) With --preview-dart-2 scroll_perf_test.dart collects timeline for a slightly longer time at https://github.com/flutter/flutter/blob/da0e898da9c329192cd1ef4f8fe6e7d03a952adc/dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart#L23-L39
On average, it has more samples compared to Dart 1 mode.

**Comment by alexmarkov** [comment]
@aam, it is still not clear to me why timeline, which is received from device has a slightly higher duration of rasterizer draw events if scroll_perf_test (which runs on host) is executed in Dart 2 mode.

**Comment by alexmarkov** [comment]
Another observation: when running host-side script in Dart 1 mode, but with increased scroll duration from 300ms to 320ms in https://github.com/flutter/flutter/blob/35b55d51a81968b381d3a1f642bc7821e7cb8d46/dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart#L30 and https://github.com/flutter/flutter/blob/35b55d51a81968b381d3a1f642bc7821e7cb8d46/dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart#L36 the benchmark regresses by ~5.8%, which is similar to the regression when it runs in Dart 2 mode.

**Comment by Hixie** [comment]
Is there a way to run the test and collect information about what the VM might be doing other than running the Dart code, e.g. in case some sort of variance in the GC timing is the cause?

**Comment by alexmarkov** [comment]
As @Hixie suggested, I experimented with adding a delay at the beginning of a scrolling test (on the host side).

It turns out that this benchmark is very sensitive to when the measurement is started.

```
Delay  |   Result (average of 10 runs, relative to plain Dart 1):

Dart 1
500ms   -5.12% (speedup!)
1500ms  +2.75%
2000ms  +4.63% (comparable to plain Dart 2)

Dart 2
500ms   +2.14% (almost no regression relative to original Dart 1)
1000ms  +7.12%
```

One of the differences between Dart 1 and Dart 2 is that Dart 2 has much slower startup (300-400ms vs 1500-1700ms for the host-side script). Slower start-up of the host side of the benchmark means that application on the device starts measurements later.

**Comment by Hixie** [comment]
I've removed the TODAY label since we now at least understand the proximate source of the regression.
I'll file separate bugs about the issues we've found here. Thanks everyone for your help.

**Comment by Hixie** [comment]
Filed https://github.com/flutter/flutter/issues/20201.

**Comment by Hixie** [comment]
I've had trouble reproducing the thing with the delays using Dart 2:

![average_frame_rasterizer_time_millis](https://user-images.githubusercontent.com/551196/43671036-39be57dc-9748-11e8-9c18-7698b8b2c401.png)

**Comment by alexmarkov** [comment]
As the benchmark is noisy, I was using an average of 10 runs.
Here is the script which runs 'flutter drive' 10 times and extracts the results:
```
#!/bin/bash
set -e
flutter clean
rm -f log1 log2
for i in {1..10}; do
  echo "========= $i ========="
  rm -f build/complex_layout_scroll_perf.timeline_summary.json
  flutter drive -v --profile --local-engine android_profile --trace-startup -t test_driver/scroll_perf.dart 2>&1 | tee log1
  cat log1 | grep average_frame_rasterizer_time_millis build/complex_layout_scroll_perf.timeline_summary.json | awk -F '[, \t]+' '{print $3}' | tee -a log2
done
echo "-----------------------------"
cat log2
```

Also note that the difference is quite small (around 5%). Given the resolution of your graph, numbers around 200-400ms delay look as non-regressed, while a couple of runs around 1000ms look as regressed.

**Comment by Hixie** [comment]
I'm still collecting data for larger numbers, but collecting 10 runs for every 20ms increment up to 540ms so far resulted in this:

![](https://user-images.githubusercontent.com/551196/43753440-acb91ee2-99ba-11e8-83ad-8f1f2fa4b707.png)

**Comment by Hixie** [comment]
Data up to 2000ms delay:
![](https://user-images.githubusercontent.com/551196/43790978-06cfdcc2-9a29-11e8-9642-78de83cd8b07.png)

**Comment by Hixie** [comment]
I'm not sure what to think. I'm not convinced this data really shows a sensitivity to an initial delay, so much as a low-frequency bad frame event.

**Comment by Hixie** [comment]
Looking at the histogram of all the data suggests that there's just a long tail of noise:

![](https://user-images.githubusercontent.com/551196/43802187-ee2316a0-9a49-11e8-8fe4-f022721d9a31.png)

**Comment by Hixie** [comment]
Looking at the median rather than the mean gives a clearer story:

![](https://user-images.githubusercontent.com/551196/43802636-3f754acc-9a4b-11e8-93a0-1846f02c395a.png)

It seems that there _is_ an initial delay sensitivity, but it's only for delays up to about 150ms.

**Comment by liyuqian** [comment]
It surprised me that the average doesn't converge well... Can we try averaging 100 data points per delay? If it's still diverging, then our benchmark's noise is so bad that it breaks the assumption of https://en.wikipedia.org/wiki/Law_of_large_numbers (which means that the variance is not finite...)

**Comment by Hixie** [comment]
Ok our conclusion is that we are spending too much time trying to track down the underlying cause here. I'm going to add a 200ms pause to get past that initial hump and then move on.

**Comment by Hixie** [comment]
@liyuqian re your last comment: the law of large numbers assumes a normal distribution. This isn't what we have here -- we have a normal distribution plus a long tail only on one side. So it's normal that the mean will be noisy. The median is much less noisy because it essentially ignores the outliers (assuming you take enough samples that less than half of them will be in the long tail part of the distribution).

The real problem with taking an average (whether a mean or a median) isn't that it doesn't converge, it's that it hides a real bug. The noise is a problem that manifests as jank to users. We shouldn't be sweeping it under the carpet, we should be figure out where it comes from. Filed https://github.com/flutter/flutter/issues/20398 to track this.

**Comment by Hixie** [comment]
Well, I guess technically it's not that the LLN assumes a normal distribution so much that with our distribution, you just need a LOT more data for the value to converge (and the expected value it will converge on isn't the peak of the curve in the chart above, it's a little to the right of the peak).

**Comment by Hixie** [comment]
Timeout added in https://github.com/flutter/flutter/pull/20400

**Comment by liyuqian** [comment]
@Hixie Yes, LLN does not assume normal distribution, it says that the mean will converge to a concentrated normal distribution with large N if the random variable has finite variance. With long tails in our random variable's distribution, it's correct that we either need a large N to see a concentrated mean, or maybe it will never converge if the tail is so long that the variance is no longer finite.

Finally, I completely agree that just looking at the mean is missing some important aspects so I'm glad that we also track 90th, 99th percentile, and the worst frame time in our dashboard. I just saw some interesting results with mean vs. 90th, 99th, and worst time when we switched clip to none by default.
